### DISCUSSION TITLE
bundling relative images with Vite doesn't seem to work?

### DISCUSSION THREAD

**Discussion by godbout** [laravel/framework#43000] [Q&A]
i'm following the docs from here: https://laravel.com/docs/9.x/vite#url-processing

my blade file is in `resources/views/index.blade.php`. my image is in `resources/img/pointer.png`. in the `index` file in point to my image as `<img src="../img/pointer.png">`. if i got the doc correctly, this should be caught by Vite and bundled. it's not. i end up with the exact same `<img src="../img/pointer.png">` in my HTML and of course the image cannot be found.

anyone has any idea? thanks.

**Comment by epcgrs** [comment]
if you are using blade, why you dont use the `asset()` method? 
ex: `asset('/img/pointer.png') `

https://laravel.com/docs/9.x/helpers#method-asset

**Comment by godbout** [comment]
hey, thanks. this is not a blocking issue. i used a static path and put my file in the public folder. the point is to understand why the new Vite system doesn't bundle images, like the documentation is describing.

**Reply by epcgrs** [reply]
ah ok, sorry but I don't have enough knowledge with Vite to answer you.

**Reply by godbout** [reply]
no worries. thanks for answering in the first place :))
yes, Vite is new. i had no idea either but it's now within the latest versions of Laravel. was starting a new site so i'm rolling with it!

**Reply by UltraWelfare** [reply]
I hit the same issue. 

As far as I understand, vite pre-processes the files defined in `defineConfig` on `vite.config.js`. So in this case, the `src="..."` is defined in the blade.php file not being processed by vite therefore not working correctly. You would need the `asset` helper in order to make it work correctly and place the image in the public path yourself.

Another (rather cumbersome) workaround would be to drop `img` element and instead use a `div` (or something else) alongside with a CSS file that would target that element with `background-image: url("...")`. That way when vite pre-processes the CSS file, it would correctly generate the URL.

**Reply by godbout** [reply]
> As far as I understand, vite pre-processes the files defined in `defineConfig` on `vite.config.js`. So in this case, the `src="..."` is defined in the blade.php file not being processed by vite therefore not working correctly. 

ah. that would make sense yeah. i didn't think too much about it. thanks for the guidance. i had added something like the image file types into Vite config but maybe it's not done for that. didn't dig too much as this is for a marketing page for a Mac app. not web dev. thanks again!

**Reply by RodgerOliver** [reply]
> I hit the same issue.
> 
> As far as I understand, vite pre-processes the files defined in `defineConfig` on `vite.config.js`. So in this case, the `src="..."` is defined in the blade.php file not being processed by vite therefore not working correctly. You would need the `asset` helper in order to make it work correctly and place the image in the public path yourself.
> 
> Another (rather cumbersome) workaround would be to drop `img` element and instead use a `div` (or something else) alongside with a CSS file that would target that element with `background-image: url("...")`. That way when vite pre-processes the CSS file, it would correctly generate the URL.

But what if the image needs to be bundled? Can Vite do that?

**Comment by lodekib** [comment]
> i'm following the docs from here: https://laravel.com/docs/9.x/vite#url-processing
> 
> my blade file is in `resources/views/index.blade.php`. my image is in `resources/img/pointer.png`. in the `index` file in point to my image as `<img src="../img/pointer.png">`. if i got the doc correctly, this should be caught by Vite and bundled. it's not. i end up with the exact same `<img src="../img/pointer.png">` in my HTML and of course the image cannot be found.
> 
> anyone has any idea? thanks.

Did you succeed? Would suggest you try `@/resources/img/pointer.png` . [check this](https://laravel-vite.dev/guide/essentials/working-with-assets.html#overview)

**Reply by godbout** [reply]
i kinda forgot about it :D i mean yes, i "succeeded" in the sense i used another way. i was just getting confused whether the doc was wrong or i was doing something wrong (latter more probable).

**Reply by P4KM4N** [reply]
It is working for me in my CSS file which is bundled by Vite, but I tried on <img> inside my blade file and it's not working for me, it is outputting the `@/images/whatever.png` and not converting it.

**Reply by RoCampos** [reply]
I'm having the same trouble with using this approach `@/resources/img/pointer.png`

**Reply by tobia** [reply]
I can only presume they've written the docs for `@/resources/img/image.png` because it was planned, but they released the Vite pipeline before that part was completed. For now I'm importing all my images in CSS or JS, because that's the only way to get hot reload in dev and versioning in build. I'm not using `public` except as a last resort.

**Comment by freemiumdev** [comment]
Hi,
I am in the same situation and the problem still remains. has anyone solved it?

**Comment by wizzymore** [comment]
Vite's URL Processing works in the files bundled by Vite ( which .blade.php files are not ).

**Reply by P4KM4N** [reply]
Make total sense! 😅 Anyway I'll probably use Glide to manipulate my image in the end, so everything is ok :)

**Reply by godbout** [reply]
i think i've seen somewhere that there's an update to some of the core npm package that takes care of stuff within Blade files too now. but haven't checked myself yet.

**Comment by adamramadhan** [comment]
it works on dev server with vite_asset helper but yeah still confused if theres automatic image detection and move it into the build/images folder too, but it seems its not working.

**Reply by innovaweb-dev** [reply]
Same problem and vite_asset not works
Call to undefined function vite_asset()
```php
<img src="{{ vite_asset('/file.png')  }}" alt="" height="16" width="16"/>
```

**Comment by innovaweb-dev** [comment]
```html
    <img src="img/sign_warning.png" alt="" height="16" width="16"/>
    <img src="resources/images/sign_warning.png" alt="" height="16" width="16"/>
    <img src="public/images/sign_warning.png" alt="" height="16" width="16"/>
    <img src="images/sign_warning.png" alt="" height="16" width="16"/>
    <img src="@/resources/img/sign_warning.png" alt="" height="16" width="16"/>
    <img src="@/resources/images/sign_warning.png" alt="" height="16" width="16"/>
    <img :src="/resources/images/sign_warning.png" alt="" height="16" width="16"/>
    <img :src="@/resources/images/sign_warning.png" alt="" height="16" width="16"/>
    <img :src="./resources/images/sign_warning.png" alt="" height="16" width="16"/>
    <img src="./resources/images/sign_warning.png" alt="" height="16" width="16"/>
    <img src="{{resource_path('images/sign_warning.png')}}" alt="" height="16" width="16"/>
    <img src="{{asset('images/sign_warning.png')}}" alt="" height="16" width="16"/>
    <img src="@vite('resources/images/sign_warning.png')" alt="" height="16" width="16"/>
    <img src="@vite('images/sign_warning.png')" alt="" height="16" width="16"/>
    <img src="@vite('sign_warning.png')" alt="" height="16" width="16"/>
```

I tried all combination as possible, nothing.
Only files which called from css or JavaScript are loaded.

A PHP framekork which need Javascript just to load images, it amazing. 
Great progress !!!

**Reply by tobia** [reply]
Yes, it does not seem to work. If you can, use CSS (background-image) which works, otherwise for the time being just put it in public.

**Reply by Uolsen** [reply]
> Yes, it does not seem to work. If you can, use CSS (background-image) which works, otherwise for the time being just put it in public.

Yeah but what if you want to use versions? And then if you would like to use the image in CSS as background and as `<img>` then there will be two different images. One for CSS “served” by vite and one copied into /public directory?

And inside the documentation of laravel (https://laravel.com/docs/9.x/vite#url-processing), there is only a showcase for the .vue file. Why? The blade is the main templating engine for Laravel. There should be a showcase of how to serve static images inside .blade.php files.

**Comment by Mohammed-Taher** [comment]
LaravelVite documentation states that the `@` symbol should resolve to `http://localhost:5173/...` but this happens currently in Javascript files only.

For `.blade.php` files I solved this issue by declaring `ASSET_URL` key in .env file:

`ASSET_URL=http://localhost:5173/resources/images`

now you can pass the image file name to `asset()` function in blade templates:

`<img src="{{ asset('image.png') }}" />`

**Reply by lukas-frey** [reply]
How will this work in production though?

**Comment by makowskid** [comment]
Solved

You have to follow this 
https://laravel.com/docs/9.x/vite#working-with-blade-and-routes

meaning, you have to add 
`import.meta.glob([
  '../images/**',
]);`
into
`resources/js/app.js`

and also make sure you have facade alias defined:
`'Vite' => \Illuminate\Support\Facades\Vite::class,`
in your `config/app.php`

then you can use 
`tyle="background-image: url('{{ Vite::asset('resources/images/photo.jpg') }}')"`
etc..

**Reply by LucasSouzaa97** [reply]
It works for me. Thank you, @makowskid akowskid!

**Reply by makowskid** [reply]
There's even a mention today about it in Laravel Newsletter
https://laravel-news.com/laravel-9-26-0

they forgot to mention about Facade alias though

**Reply by sebsobseb** [reply]
The `Vite` Facade alias is automatically available normally (9.x), didn't had to do it

**Reply by makowskid** [reply]
wasn't available earlier :)

**Reply by ShellarFX** [reply]
Hello. I tried this, but i have an error
```
Could not resolve entry module (resources/images/*).
error during build:
Error: Could not resolve entry module (resources/images/*).
    at error (file:///Users/admin/PhpstormProjects/market/node_modules/rollup/dist/es/shared/rollup.js:1858:30)
    at ModuleLoader.loadEntryModule (file:///Users/admin/PhpstormProjects/market/node_modules/rollup/dist/es/shared/rollup.js:22369:20)
    at async Promise.all (index 6)
```

**Reply by shinnida220** [reply]
Thanks man, you are a life saver. Worked like magic

**Reply by makowskid** [reply]
@ShellarFX maybe try with `/resources/images/*` ?

**Reply by leandrov27** [reply]
/resources/images/**

**Reply by makowskid** [reply]
sorry, correct :)
